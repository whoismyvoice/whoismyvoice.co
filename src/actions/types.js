// @flow

/** Action type used when address is provided. */
export const RECEIVE_ADDRESS = 'RECEIVE_ADDRESS';
/** Action type used when new contribution data is received. */
export const RECEIVE_CONTRIBUTION_DATA = 'RECEIVE_CONTRIBUTION_DATA';
/** Action type used when gps or location data is provided. */
export const RECEIVE_GPS = 'RECEIVE_GPS';
/** Action type used after officials are updated. */
export const RECEIVE_OFFICIALS = 'RECEIVE_OFFICIALS';
/** Action type used when receiving full list of legislators. */
export const RECEIVE_OFFICIALS_ALL = 'RECEIVE_OFFICIALS_ALL';
/** Action type used when receiving an error after requesting officials for address. */
export const RECEIVE_OFFICIALS_ERROR = 'RECEIVE_OFFICIALS_ERROR';
/** Action type used when address is provided. */
export const RECEIVE_ZIP_CODE = 'RECEIVE_ZIP_CODE';
/** Action type to reset current state. */
export const RESET_CURRENT = 'RESET_CURRENT';
/** Action to toggle the menu. */
export const TOGGLE_MENU = 'TOGGLE_MENU';

/**
 * Record type for an Action.
 */
export type Action = {
  type: string,
  [key: string]: any,
};
