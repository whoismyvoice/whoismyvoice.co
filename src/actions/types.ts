import { Dispatch as ReduxDispatch } from 'redux';
import { Contribution, SectorContributions } from '../models/Contribution';
import {
  CongressPerson,
  Record as Legislator,
  Senator,
} from '../models/Legislator';
import { Office } from '../models/Office';

export enum ActionType {
  /** Action type used when address is provided. */
  RECEIVE_ADDRESS = 'RECEIVE_ADDRESS',
  /** Action type used when new contribution data is received. */
  RECEIVE_CONTRIBUTION_DATA = 'RECEIVE_CONTRIBUTION_DATA',
  /** Action type used when data for multiple contributions is received. */
  RECEIVE_CONTRIBUTIONS_BY_SECTOR_DATA = 'RECEIVE_CONTRIBUTIONS_BY_SECTOR_DATA',
  /** Action type used when data for multiple contributions is received. */
  RECEIVE_CONTRIBUTIONS_DATA = 'RECEIVE_CONTRIBUTIONS_DATA',
  /** Action type used when gps or location data is provided. */
  RECEIVE_GPS = 'RECEIVE_GPS',
  /** Action type dispatched when full house retrieved. */
  RECEIVE_HOUSE = 'RECEIVE_HOUSE',
  /** Action type dispatched when office divisions retrieved. */
  RECEIVE_OFFICES = 'RECEIVE_OFFICES',
  /** Action type used when receiving full list of legislators. */
  RECEIVE_OFFICIALS_ALL = 'RECEIVE_OFFICIALS_ALL',
  /** Action type used when receiving an error after requesting officials for address. */
  RECEIVE_OFFICES_ERROR = 'RECEIVE_OFFICES_ERROR',
  /** Action type dispatched when change pages of legislators. */
  RECEIVE_PAGE = 'RECEIVE_PAGE',
  /** Action type dispatched when full senate retrieved. */
  RECEIVE_SENATE = 'RECEIVE_SENATE',
  /** Action type used when address is provided. */
  RECEIVE_ZIP_CODE = 'RECEIVE_ZIP_CODE',
  /** Action type used when invalid zip code is provided. */
  RECEIVE_ZIP_CODE_INVALID = 'RECEIVE_ZIP_CODE_INVALID',
  /** Action type to reset current state. */
  RESET_CURRENT = 'RESET_CURRENT',
  /** Action to toggle the menu. */
  TOGGLE_MENU = 'TOGGLE_MENU',
  /** Catch all action type */
  OTHER = '__any_other_action__',
}

export interface AddressAction {
  type: ActionType.RECEIVE_ADDRESS;
  address: string;
}

export interface ContributionDataAction {
  type: ActionType.RECEIVE_CONTRIBUTION_DATA;
  amount: number;
  legislatorId: string;
  organization: string;
}

export interface ContributionsDataAction {
  type: ActionType.RECEIVE_CONTRIBUTIONS_DATA;
  contributions: Array<Contribution>;
}

export interface ContributionsBySectorDataAction {
  type: ActionType.RECEIVE_CONTRIBUTIONS_BY_SECTOR_DATA;
  contributions: SectorContributions[];
}

export interface HouseAction {
  type: ActionType.RECEIVE_HOUSE;
  congressPersons: CongressPerson[];
}

export interface OfficesAction {
  type: ActionType.RECEIVE_OFFICES;
  offices: Office[];
}

export interface LegislatorsAction {
  type: ActionType.RECEIVE_OFFICIALS_ALL;
  officials: Array<Legislator>;
}

export interface OfficesErrorAction {
  type: ActionType.RECEIVE_OFFICES_ERROR;
  code: number;
  isGlobal: boolean;
  isParseError: boolean;
  message: string;
  messages: Array<string>;
}

export interface SenateAction {
  type: ActionType.RECEIVE_SENATE;
  senators: Senator[];
}

export interface ZipCodeAction {
  type: ActionType.RECEIVE_ZIP_CODE;
  zipCode: string;
}

export interface ZipCodeInvalidAction {
  type: ActionType.RECEIVE_ZIP_CODE_INVALID;
}

export interface ResetAction {
  type: ActionType.RESET_CURRENT;
}

export interface ToggleMenuAction {
  type: ActionType.TOGGLE_MENU;
}

export interface OtherAction {
  type: ActionType.OTHER;
}

export interface PageAction {
  type: ActionType.RECEIVE_PAGE;
  page: 1 | 2;
}

export interface GpsAction {
  type: ActionType.RECEIVE_GPS;
  latitude: string;
  longitude: string;
}

/**
 * Record type for an Action.
 */
export type Action =
  | AddressAction
  | ContributionDataAction
  | ContributionsDataAction
  | ContributionsBySectorDataAction
  | HouseAction
  | LegislatorsAction
  | OfficesAction
  | OfficesErrorAction
  | ZipCodeAction
  | ZipCodeInvalidAction
  | ResetAction
  | SenateAction
  | ToggleMenuAction
  | GpsAction
  | PageAction
  | OtherAction;

/**
 * Type for the redux dispatch function.
 */
export type Dispatch = ReduxDispatch<Action>;
