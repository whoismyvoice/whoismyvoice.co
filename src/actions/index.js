import PropTypes from 'prop-types';
import fetch from 'isomorphic-fetch';

import {
  ELECTION_CYCLE,
  EXECUTE_PROXY,
  ORGANIZATION,
} from '../constants';
import {
  RECEIVE_ADDRESS,
  RECEIVE_CONTRIBUTION_DATA,
  RECEIVE_OFFICIALS,
  RECEIVE_OFFICIALS_ALL,
  RECEIVE_ZIP_CODE,
  RESET_CURRENT,
  TOGGLE_MENU,
} from './types';
import { Legislator, } from '../models/Legislator';
import { PropType as MaplightResultsType, } from '../models/MaplightResults';

/**
 * Match FEC ids for house and senate races. Potentially important because some
 * of these candidates may run for other offices (e.g. President) and we are not
 * interested in those contributions.
 */
const FEC_ID_REGEX = /[HS]\d{1,3}[A-Z]{2}\d+/;

/**
 * Retrieve aggregated contribution data for given `organization` and `legislator`.
 * @param {string} organization name whose contributions will be searched.
 * @param {object} legislator for whom data will be retrieved.
 * @param {object} legislator.id object of identifiers for the `legislator`.
 * @param {array} legislator.id.fec string identifiers for FEC filings.
 * @returns contribution data.
 */
async function fetchContributionsForCandidate(organization, legislator) {
  const candidateFecIds = legislator.id.fec.filter(fecId => FEC_ID_REGEX.test(fecId));
  const baseUrl = `${EXECUTE_PROXY}/maplight`;
  const electionCycle = encodeURIComponent(ELECTION_CYCLE);
  const organizationName = encodeURIComponent(organization);
  const fecIds = encodeURIComponent(candidateFecIds.join('|'));
  const params = `cycle=${electionCycle}&fecIds=${fecIds}&organization=${organizationName}`;
  const url = `${baseUrl}?${params}`;
  const response = await fetch(url, {
    mode: 'cors',
    headers: {
      'Content-Type': 'text/plain',
    },
  });
  if (response.ok) {
    const body = await response.json();
    return body;
  } else {
    const error = new Error(response.statusText);
    error.response = response;
    throw error;
  }
}

/**
 * Retrieve data about officials/representatives based the provided address.
 * @param {string} address to use in lookup.
 * @returns array of officials.
 */
async function fetchOfficialsForAddress(address) {
  const baseUrl = `${EXECUTE_PROXY}/civics`;
  const encodedAddress = encodeURIComponent(address);
  const params = `address=${encodedAddress}`;
  const response = await fetch(`${baseUrl}?${params}`);
  if (response.ok) {
    const body = await response.json();
    const officials = body.officials;
    return officials;
  } else {
    const error = new Error(response.statusText);
    error.response = response;
    throw error;
  }
}

/**
 * Retreive data about officials/legislators for the current legislative session.
 * @returns array of officials.
 */
async function fetchLegislatorsAll() {
  const url = 'https://theunitedstates.io/congress-legislators/legislators-current.json';
  const response = await fetch(url);
  if (response.ok) {
    const body = await response.json();
    const officials = body;
    return officials;
  } else {
    const error = new Error(response.statusText);
    error.response = response;
    throw error;
  }
}

async function getLegislatorForOfficial(allLegislators, official) {
  return allLegislators.find((legislator) => legislator.name.official_full === official.name);
}

/**
 * Create action to notify address has been set.
 * @param {array} address received.
 * @returns action.
 */
export function receiveAddress(address) {
  return {
    type: RECEIVE_ADDRESS,
    address,
  };
}

/**
 * Create action to notify of contribution data received.
 * @param {object} legislator for whom data was received.
 * @param {object} legislator.name object of name information for the `legislator`.
 * @param {array} legislator.name.official_full official full name of the legislator
 *    according to FEC records.
 * @param {object} contributionResults received from Maplight API
 * @param {object} contributionData.search_terms information about what was searched.
 * @param {object} contributionData.search_terms.donor information about the donor
 *    search parameters
 * @param {object} contributionData.search_terms.donor.donor_organization
 * @param {object} contributionData.data result data.
 * @param {array} contributionData.data.aggregate_totals grouped by ????. Each member
 *    has a `total_amount` property that is the aggregated value of contributions.
 * @returns action.
 */
function receiveContributionDataForLegislator(legislator, contributionResults) {
  PropTypes.checkPropTypes(
    { contributionResults: MaplightResultsType, },
    { contributionResults, },
    'contributionResults',
    'actions#receiveContributionDataForLegislator',
  );
  return receiveContributionData(
    Legislator.getIdentifier(legislator),
    contributionResults.search_terms.donor.donor_organization,
    contributionResults.data.aggregate_totals[0].total_amount,
  );
}

/**
 * Create action to notify of contribution data received.
 * @param {string} legislatorId official full name of the legislator
 *    according to FEC records.
 * @param {string} organization making the contribution.
 * @param {number} amount received in contributions.
 * @returns action.
 */
export function receiveContributionData(legislatorId, organization, amount) {
  return {
    type: RECEIVE_CONTRIBUTION_DATA,
    amount,
    legislatorId,
    organization,
  };
}

/**
 * Create action to notify officials retrieved.
 * @param {array} officials received.
 * @returns action.
 */
export function receiveOfficials(officials) {
  return {
    type: RECEIVE_OFFICIALS,
    officials,
  };
}

/**
 * Create action to notify officials retrieved.
 * @param {array} officials received.
 * @returns action.
 */
export function receiveOfficialsAll(officials) {
  return {
    type: RECEIVE_OFFICIALS_ALL,
    officials,
  };
}

/**
 * Create action to notify address has been set.
 * @param {array} address received.
 * @returns action.
 */
export function receiveZipCode(zipCode) {
  return {
    type: RECEIVE_ZIP_CODE,
    zipCode,
  };
}

/**
 * Create action to reset current state.
 * @returns action.
 */
export function reset() {
  return {
    type: RESET_CURRENT,
  };
}

/**
 * Create a TOGGLE_MENU action instance.
 * @returns action.
 */
export function toggleMenu() {
  return {
    type: TOGGLE_MENU,
  };
}

/**
 * Asynchronously dispatch updates to zip code and address.
 */
export function setZipCode(zipCode) {
  return async dispatch => {
    dispatch(receiveZipCode(zipCode));
    setAddress(zipCode)(dispatch);
  };
}

/**
 * Asynchronously dispatch updates to address.
 */
export function setAddress(address) {
  return async dispatch => {
    dispatch(receiveAddress(address));
    const allLegislators = await fetchLegislatorsAll();
    const currentOfficials = await fetchOfficialsForAddress(address);
    dispatch(receiveOfficialsAll(allLegislators));
    dispatch(receiveOfficials(currentOfficials));
    const getLegislator = getLegislatorForOfficial.bind(this, allLegislators);
    const currentLegislators = await Promise.all(currentOfficials.map(official => getLegislator(official)));
    currentLegislators.forEach(async legislator => {
      const contributionData = await fetchContributionsForCandidate(ORGANIZATION, legislator);
      dispatch(receiveContributionDataForLegislator(legislator, contributionData));
    });
  };
}
