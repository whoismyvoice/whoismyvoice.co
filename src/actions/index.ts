import unfetch from 'isomorphic-unfetch';
import * as mixpanel from 'mixpanel-browser';

import { ELECTION_CYCLE, ORGANIZATION } from '../constants';
import { Action, ActionType, Dispatch } from './types';
import {
  Contribution,
  SectorContribution,
  SectorContributions,
} from '../models/Contribution';
import {
  Legislator,
  Identifier as LegislatorIdentifier,
  Record as LegislatorRecord,
} from '../models/Legislator';
import { MaplightResultsRecord } from '../models/MaplightResults';
import { Record as Official } from '../models/Official';
import { GoogleResponseError } from '../models/GoogleResponseError';
import { ResponseError } from '../models/ResponseError';

const fetch = unfetch;

/**
 * Match FEC ids for house and senate races. Potentially important because some
 * of these candidates may run for other offices (e.g. President) and we are not
 * interested in those contributions.
 */
const FEC_ID_REGEX = /[HS]\d{1,3}[A-Z]{2}\d+/;

/**
 * Type guard to check if the given item exists (is not `null` and not
 * `undefined`.
 * @param item to test.
 * @return `true` iff `item` is neither `null` nor `undefined`.
 */
function isDefined<T>(item: T | undefined | null): item is T {
  return item !== undefined && item !== null;
}

/**
 * Finesse the `MaplightResultsRecord` and `LegislatorRecord` data into a
 * `Contribution`.
 * @param {LegislatorRecord} legislator from which identifier will be retrieved.
 * @param {MaplightResultsRecord} contributionData from which amount and
 *    organization will be retrieved.
 * @returns the data as a `Contribution`.
 */
function createContribution(
  legislator: LegislatorRecord,
  contributionData: MaplightResultsRecord
): Contribution {
  return {
    legislatorId: Legislator.getIdentifier(legislator),
    organization: contributionData.search_terms.donor.donor_organization,
    amount: contributionData.data.aggregate_totals[0].total_amount,
  };
}

/**
 * Create a `SectorContribution` from the given `node` by reading its properties.
 * @param node from which sector contribution data will be parsed.
 * @returns a record with the sector contribution information if it could be
 * parsed, `undefined` if it could not be parsed from the attributes of `node`.
 */
function createSectorContribution(
  node: Element
): SectorContribution | undefined {
  const sector = node.getAttribute('sector_name');
  const sectorCode = node.getAttribute('sectorid');
  const total = node.getAttribute('total');
  if (sector === null || sectorCode === null || total === null) {
    return undefined;
  }
  const amount = parseInt(total, 10);
  if (isNaN(amount)) {
    return undefined;
  } else {
    return {
      amount,
      sector,
      sectorCode,
    };
  }
}

/**
 * Finesse the `Document` from the Open Secrets API into a list of
 * contributions.
 * @param legislator from which identifier will be retrieved.
 * @param document from which sector contribution data will be parsed.
 * @returns a record with the legislator id and a list of sector contribution
 * data.
 */
function createSectorContributions(
  legislator: LegislatorRecord,
  document: Document
): SectorContributions {
  const nodes = document.getElementsByTagName('sector');
  const legislatorId = Legislator.getIdentifier(legislator);
  const elements = Array.from(nodes);
  const contributions = elements
    .map(createSectorContribution)
    .filter(isDefined);
  return {
    legislatorId,
    contributions,
  };
}

/**
 * Get contributions to the given `legislator` from the `organization`.
 * @param organization whose contributions to `legislator` should be retrieved.
 * @param legislator for whome `organization` contributions will be retrieved.
 * @returns a `Contribution` record for `legislator` from `organization`.
 */
async function fetchContributionByOrg(
  organization: string,
  legislator: LegislatorRecord
): Promise<Contribution> {
  const getContributions = fetchContributions(organization);
  const contributionData = await getContributions(legislator);
  return createContribution(legislator, contributionData);
}

/**
 * Retrieve aggregated contribution data for given `organization` and `legislator`.
 * @param {string} organization name whose contributions will be searched.
 * @param {LegislatorRecord} legislator for whom data will be retrieved.
 * @returns contribution data.
 */
function fetchContributions(organization: string) {
  return async (
    legislator: LegislatorRecord
  ): Promise<MaplightResultsRecord> => {
    const candidateFecIds = legislator.id.fec.filter((fecId) =>
      FEC_ID_REGEX.test(fecId)
    );
    const baseUrl = '/api/contributions';
    const electionCycle = encodeURIComponent(await ELECTION_CYCLE);
    const organizationName = encodeURIComponent(organization);
    const fecIds = encodeURIComponent(candidateFecIds.join('|'));
    const params = `cycle=${electionCycle}&fecIds=${fecIds}&organization=${organizationName}`;
    const url = `${baseUrl}?${params}`;
    const response = await fetch(url, {
      mode: 'cors',
      headers: {
        'Content-Type': 'text/plain',
      },
    });
    if (response.ok) {
      const body = await response.json();
      return body as MaplightResultsRecord;
    } else {
      throw new ResponseError(response, response.statusText);
    }
  };
}

/**
 * Retrieve contributions group by business sector.
 * @param {LegislatorRecord} legislator for whom data will be retrieved.
 * @returns contribution data.
 */
async function fetchContributionsBySector(
  legislator: LegislatorRecord
): Promise<SectorContributions> {
  const opensecretsId = legislator.id.opensecrets;
  const baseUrl = '/api/contributions/by-sector';
  const electionCycle = await ELECTION_CYCLE;
  const encodedCycle = encodeURIComponent(electionCycle.split('|')[0]);
  const encodedId = encodeURIComponent(opensecretsId);
  const params = `cycle=${encodedCycle}&id=${encodedId}`;
  const url = `${baseUrl}?${params}`;
  const response = await fetch(url, {
    mode: 'cors',
    headers: {
      'Content-Type': 'text/xml',
    },
  });
  if (response.ok) {
    const body = await response.text();
    // Parse XML document
    const parser = new DOMParser();
    const doc = parser.parseFromString(body, 'text/xml');
    return createSectorContributions(legislator, doc);
  } else {
    throw new ResponseError(response, response.statusText);
  }
}

/**
 * Retrieve data about officials/representatives based the provided address.
 * @param {string} address to use in lookup.
 * @returns array of officials.
 */
async function fetchOfficialsForAddress(address: string): Promise<Official[]> {
  const baseUrl = `/api/civic-information`;
  const encodedAddress = encodeURIComponent(address);
  const params = `address=${encodedAddress}`;
  const response = await fetch(`${baseUrl}?${params}`);
  if (response.ok) {
    const body = await response.json();
    const officials = body.officials;
    return officials as Array<Official>;
  } else {
    throw new GoogleResponseError(response, response.statusText);
  }
}

/**
 * Retreive data about officials/legislators for the current legislative session.
 * @returns array of officials.
 */
async function fetchLegislatorsAll(): Promise<LegislatorRecord[]> {
  const url =
    'https://theunitedstates.io/congress-legislators/legislators-current.json';
  const response = await fetch(url);
  if (response.ok) {
    return (await response.json()) as Array<LegislatorRecord>;
  } else {
    throw new ResponseError(response, response.statusText);
  }
}

/**
 * Create a function that searches `allLegislators` for `official`.
 * @param {Array<LegislatorRecord>} allLegislators records to be searched.
 * @returns function that tries to match given `official` to a
 *    `LegislatorRecord`.
 */
function getLegislatorForOfficial(allLegislators: Array<LegislatorRecord>) {
  return (official: Official): LegislatorRecord | undefined =>
    allLegislators.find(
      (legislator) =>
        Legislator.getIdentifier(legislator) ===
        Legislator.getIdentifier(official)
    );
}

/**
 * Create action to navigate to the specified group of legislators.
 * @param {number} page indicate group of legislators to view.
 * @returns action.
 */
export function navigateLegislators(page: 1 | 2): Action {
  return {
    type: ActionType.RECEIVE_PAGE,
    page,
  };
}

/**
 * Create action to notify address has been set.
 * @param {array} address received.
 * @returns action.
 */
export function receiveAddress(address: string): Action {
  mixpanel.track(ActionType.RECEIVE_ADDRESS);
  return {
    type: ActionType.RECEIVE_ADDRESS,
    address,
  };
}

/**
 * Create action to notify of contribution data received.
 * @param {string} legislatorId official full name of the legislator
 *    according to FEC records.
 * @param {string} organization making the contribution.
 * @param {number} amount received in contributions.
 * @returns action.
 */
export function receiveContribution(
  legislatorId: LegislatorIdentifier,
  organization: string,
  amount: number
): Action {
  return {
    type: ActionType.RECEIVE_CONTRIBUTION_DATA,
    amount,
    legislatorId,
    organization,
  };
}

/**
 * Create action to notify of bulk contribution data received.
 * @param {Array<Contribution>} contributions received.
 * @returns action.
 */
export function receiveContributions(
  contributions: Array<Contribution>
): Action {
  mixpanel.track(ActionType.RECEIVE_CONTRIBUTION_DATA, {
    count: contributions.length,
  });
  return {
    type: ActionType.RECEIVE_CONTRIBUTIONS_DATA,
    contributions: contributions,
  };
}

/**
 * Create action to notify of bulk contribution sector data received.
 * @param contributions received.
 * @returns action.
 */
export function receiveContributionsBySector(
  contributions: SectorContributions[]
): Action {
  mixpanel.track(ActionType.RECEIVE_CONTRIBUTION_DATA, {
    count: contributions.length,
  });
  return {
    type: ActionType.RECEIVE_CONTRIBUTIONS_BY_SECTOR_DATA,
    contributions: contributions,
  };
}

/**
 * Create action to notify officials retrieved.
 * @param {array} officials received.
 * @returns action.
 */
export function receiveOfficials(officials: Array<Official>): Action {
  mixpanel.track(ActionType.RECEIVE_OFFICIALS, { count: officials.length });
  return {
    type: ActionType.RECEIVE_OFFICIALS,
    officials,
  };
}

/**
 * Create action to notify officials retrieved.
 * @param {array} officials received.
 * @returns action.
 */
export function receiveOfficialsAll(
  officials: Array<LegislatorRecord>
): Action {
  mixpanel.track(ActionType.RECEIVE_OFFICIALS_ALL);
  return {
    type: ActionType.RECEIVE_OFFICIALS_ALL,
    officials,
  };
}

/**
 * Create action to notify officials retrieved an error.
 * @param {error} error received.
 * @returns action.
 */
export async function receiveOfficialsError(
  error: GoogleResponseError
): Promise<Action> {
  mixpanel.track(ActionType.RECEIVE_OFFICIALS_ERROR, {
    message: error.message,
  });
  return {
    type: ActionType.RECEIVE_OFFICIALS_ERROR,
    code: await error.code,
    isGlobal: await error.isGlobal,
    isParseError: await error.isParseError,
    message: await error.responseMessage,
    messages: await error.errorMessages,
  };
}

/**
 * Create action to notify address has been set.
 * @param {array} address received.
 * @returns action.
 */
export function receiveZipCode(zipCode: string): Action {
  return {
    type: ActionType.RECEIVE_ZIP_CODE,
    zipCode,
  };
}

/**
 * Create an invalid zip code action.
 * @returns an action indicating zip code is not valid.
 */
export function receiveZipCodeInvalid(): Action {
  return {
    type: ActionType.RECEIVE_ZIP_CODE_INVALID,
  };
}

/**
 * Create action to reset current state.
 * @returns action.
 */
export function reset(): Action {
  mixpanel.track(ActionType.RESET_CURRENT);
  return {
    type: ActionType.RESET_CURRENT,
  };
}

/**
 * Create a TOGGLE_MENU action instance.
 * @returns action.
 */
export function toggleMenu(): Action {
  mixpanel.track(ActionType.TOGGLE_MENU);
  return {
    type: ActionType.TOGGLE_MENU,
  };
}

/**
 * Asynchronously dispatch updates to zip code and address.
 */
export function setZipCode(zipCode: string) {
  return (dispatch: Dispatch): void => {
    dispatch(receiveZipCode(zipCode));
    setAddress(zipCode)(dispatch);
  };
}

/**
 * Asynchronously dispatch updates to address.
 */
export function setAddress(address: string) {
  return async (dispatch: Dispatch): Promise<void> => {
    try {
      dispatch(receiveAddress(address));
      const [allLegislators, officials] = await Promise.all([
        fetchLegislatorsAll(),
        fetchOfficialsForAddress(address),
      ]);
      dispatch(receiveOfficialsAll(allLegislators));
      dispatch(receiveOfficials(officials));
      const fetchContribution = fetchContributionByOrg.bind(null, ORGANIZATION);
      const getLegislator = getLegislatorForOfficial(allLegislators);
      const legislators = officials.map(getLegislator).filter(isDefined);
      Promise.all(
        legislators.map((record) => fetchContributionsBySector(record))
      )
        .then(receiveContributionsBySector)
        .then(dispatch);
      Promise.all(legislators.map(fetchContribution))
        .then(receiveContributions)
        .then(dispatch);
    } catch (error) {
      if (error instanceof GoogleResponseError) {
        // response is error; abort
        receiveOfficialsError(error).then(dispatch);
        return;
      }
    }
  };
}
