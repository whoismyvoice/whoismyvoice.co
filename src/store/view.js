// @flow

import {
  RECEIVE_OFFICIALS,
  RECEIVE_OFFICIALS_ERROR,
  TOGGLE_MENU,
} from '../actions/types';

import type { Action } from '../actions/types';

type UninitializedString = string | void;

export type ViewState = {
  addressErrorMessage: UninitializedString,
  isMenuOpen: boolean,
};

const initialState = {
  addressErrorMessage: undefined,
  isMenuOpen: false,
};

function handle(state: ViewState = initialState, action: Action): ViewState {
  const { type } = action;
  switch (type) {
    case RECEIVE_OFFICIALS:
      return {
        ...state,
        addressErrorMessage: undefined,
      };
    case RECEIVE_OFFICIALS_ERROR:
      return {
        ...state,
        addressErrorMessage: action.error.message,
      };
    case TOGGLE_MENU:
      return {
        ...state,
        isMenuOpen: !state.isMenuOpen,
      };
    default:
      return state;
  }
}

export default handle;
