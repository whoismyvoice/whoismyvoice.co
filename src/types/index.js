import PropTypes from 'prop-types';

export const legislatorType = PropTypes.arrayOf(PropTypes.shape({
  id: PropTypes.shape({
    bioguide: PropTypes.string,
    fec: PropTypes.arrayOf(PropTypes.string),
  }),
  name: PropTypes.shape({
    first: PropTypes.string,
    last: PropTypes.string,
    official_full: PropTypes.string,
  }),
  bio: PropTypes.shape({
    birthday: PropTypes.string,
    gender: PropTypes.string,
  }),
  terms: PropTypes.arrayOf(PropTypes.shape({
    type: PropTypes.string,
    start: PropTypes.string,
    end: PropTypes.string,
    state: PropTypes.string,
    district: PropTypes.number,
    party: PropTypes.string,
  })),
}));