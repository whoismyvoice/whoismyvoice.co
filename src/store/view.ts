import { Action, ActionType } from '../actions/types';

export type ViewState = {
  addressErrorMessage?: string;
  isMenuOpen: boolean;
};

const initialState = {
  addressErrorMessage: undefined,
  isMenuOpen: false,
};

function handle(state: ViewState = initialState, action: Action): ViewState {
  switch (action.type) {
    case ActionType.RECEIVE_OFFICIALS:
      return {
        ...state,
        addressErrorMessage: undefined,
      };
    case ActionType.RECEIVE_OFFICIALS_ERROR:
      return {
        ...state,
        addressErrorMessage: action.error.message,
      };
    case ActionType.TOGGLE_MENU:
      return {
        ...state,
        isMenuOpen: !state.isMenuOpen,
      };
    default:
      return state;
  }
}

export default handle;
