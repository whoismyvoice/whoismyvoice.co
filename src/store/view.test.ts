import {
  receiveOfficesError,
  toggleMenu,
  receiveAddress,
  receiveOffices,
  receiveZipCode,
  reset,
  navigateLegislators,
  receiveZipCodeInvalid,
} from '../actions';
import { Action, ActionType } from '../actions/types';
import { GoogleResponseError } from '../models/GoogleResponseError';
import view, { ViewState } from './view';

function action(): Action {
  return {
    type: ActionType.OTHER,
  };
}

function createResponse(): Response {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-expect-error not a full Response object
  return {
    json: () => {
      return Promise.resolve({
        error: { message: 'foo' },
      });
    },
    ok: true,
    status: 200,
  };
}

describe('initial state', () => {
  const initialState = view(undefined, action());
  it('provides an initial state when given undefined', () => {
    expect(initialState).toBeDefined();
  });
  it('has the menu closed in initial state', () => {
    expect(initialState.isMenuOpen).toBe(false);
  });
  it('has an undefined address error message', () => {
    expect(initialState.addressErrorMessage).toBe(undefined);
  });
  it('has an undefined current page', () => {
    expect(initialState.currentPage).toBeUndefined();
  });

  describe('toggleMenu', () => {
    const state = view(initialState, toggleMenu());
    it('opens the menu', () => {
      expect(state.isMenuOpen).toBe(true);
    });
  });

  describe('receiveOfficialsError', () => {
    let receiveOfficialsErrorAction = undefined;
    let state = initialState;
    const error = new GoogleResponseError(createResponse());
    beforeEach(async () => {
      receiveOfficialsErrorAction = await receiveOfficesError(error);
      state = view(initialState, receiveOfficialsErrorAction);
    });
    it('has an error', () => {
      expect(state.addressErrorMessage).toBe('foo');
    });

    describe('receiveOffices', () => {
      let newState = view(state, receiveOffices([]));
      beforeEach(() => {
        newState = view(state, receiveOffices([]));
      });
      it('no longer has an error', () => {
        expect(newState.addressErrorMessage).toBe(undefined);
      });
    });

    describe('reset', () => {
      let newState = view(state, reset());
      beforeEach(() => {
        newState = view(state, reset());
      });
      it('no longer has an error', () => {
        expect(newState.addressErrorMessage).toBe(undefined);
      });
    });
  });

  describe('receiveOffices', () => {
    let state = view(initialState, receiveOffices([]));
    beforeEach(() => {
      state = view(initialState, receiveOffices([]));
    });
    it('does not change address error message', () => {
      expect(state.addressErrorMessage).toBe(undefined);
    });
  });

  describe('navigateLegislators(1)', () => {
    let state = view(initialState, navigateLegislators(1));
    beforeEach(() => {
      state = view(initialState, navigateLegislators(1));
    });
    it('changes current page', () => {
      expect(state.currentPage).toBe(1);
    });

    describe('reset', () => {
      let newState = view(state, reset());
      beforeEach(() => {
        newState = view(state, reset());
      });
      it('no longer has an error', () => {
        expect(newState.addressErrorMessage).toBe(undefined);
      });
    });

    describe('navigateLegislators(2)', () => {
      let newState = view(initialState, navigateLegislators(2));
      beforeEach(() => {
        newState = view(initialState, navigateLegislators(2));
      });
      it('changes current page', () => {
        expect(newState.currentPage).toBe(2);
      });

      describe('reset', () => {
        let nextState = view(newState, reset());
        beforeEach(() => {
          nextState = view(newState, reset());
        });
        it('no longer has an error', () => {
          expect(nextState.addressErrorMessage).toBe(undefined);
        });
      });
    });
  });

  describe('ZipCodeInvalid', () => {
    let state: ViewState;
    beforeEach(() => {
      state = view(initialState, receiveZipCodeInvalid());
    });
    it('has an address error message', () => {
      expect(state.addressErrorMessage).toBeDefined();
    });
    it('has an address error message string', () => {
      expect(state.addressErrorMessage).toMatchSnapshot();
    });
    describe('receiveZipCode', () => {
      beforeEach(() => {
        state = view(state, receiveZipCode('12345'));
      });
      it('has address error message cleared', () => {
        expect(state.addressErrorMessage).toBeUndefined();
      });
    });
    describe('receiveAddress', () => {
      beforeEach(() => {
        state = view(state, receiveAddress('12345'));
      });
      it('has address error message cleared', () => {
        expect(state.addressErrorMessage).toBeUndefined();
      });
    });
  });
});
