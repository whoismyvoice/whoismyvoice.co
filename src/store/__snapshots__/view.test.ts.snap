// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initial state ZipCodeInvalid has an address error message string 1`] = `"Valid zip code is required."`;
